src/rt/ds/stack.h

```diff
@@ -7,128 +7,302 @@
     {
       assert(!empty());
+      return !is_empty(index - 1);
+    }
 
+    /// Call this to pop an element from the stack.  Only
+    /// correct to call this if pop_is_fast just returned
+    /// true.
+    ALWAYSINLINE T* pop_fast()
+    {
+      assert(pop_is_fast());
+      auto item = peek();
       index--;
-      T item = block->data[index];
+      return item;
+    }
 
-      if (index == 0)
-        pop_slow_path();
+    /// Call this to pop an element from the stack.  Only
+    /// correct to call this if pop_is_fast just returned
+    /// false.  This returns a pair of the popped element
+    /// and the block that the client must dispose of.
+    std::pair<T*, Block*> pop_slow()
+    {
+      auto item = peek();
+      T** prev_index = get_block(index)->prev;
+      auto dealloc = get_block(index);
+      index = prev_index;
+      return {item, dealloc};
+    }
 
-      return item;
+    /// Call this to determine if push can proceed without
+    /// a new block.
+    ALWAYSINLINE bool push_is_fast()
+    {
+      return is_not_full(index);
     }
 
-    ALWAYSINLINE void push(T item)
+    /// Call this to push an element onto the stack.  Only
+    /// correct to call this if push_is_fast just returned
+    /// true.
+    ALWAYSINLINE void push_fast(T* item)
     {
-      if (index < STACK_COUNT)
-      {
-        block->data[index] = item;
-        index++;
-      }
-      else
-      {
-        push_slow_path(item);
-      }
+      assert(push_is_fast());
+      index++;
+      *index = item;
     }
 
-    template<void apply(T t)>
+    /// Call this to push an element onto the stack.  Only
+    /// correct to call this if push_is_fast just returned
+    /// false.  It needs to be provided a new block of memory
+    /// for the stack to use.
+    void push_slow(T* item, Block* block)
+    {
+      assert(!push_is_fast());
+
+      T** iter = (T**)block;
+      assert(is_empty(iter));
+      auto next = get_block(iter);
+
+      assert(index != (T**)&null_block);
+      next->prev = index;
+      index = &(next->data[0]);
+      next->data[0] = item;
+    }
+
+    /// For all elements of the stack
+    template<void apply(T* t)>
     void forall()
     {
-      Block* curr = block;
-      size_t i = index;
+      T* curr = index;
 
-      while (curr != nullptr)
+      while (curr != null_index)
       {
         do
         {
-          i--;
-          apply(curr->data[i]);
-        } while (i > 0);
+          apply(*curr);
+          curr--;
+        } while (is_empty(curr));
 
-        curr = curr->prev;
-        i = STACK_COUNT;
+        curr = get_block(curr)->prev;
       }
     }
+  };
 
-  private:
-    void pop_slow_path()
+  /**
+   * This class uses the block structured stack with extra fields
+   * for the allocator, and a backup block, so that the common case
+   * of 0-1 elements can be fast, and any other block boundrary case.
+   */
+  template<class T, class Alloc>
+  class Stack
```

Comment:
The logic for `Stack` and `StackSmall` is very similar.  Would `Stack` be simpler if it had an inner class that wrapped an `Alloc` and managed the `backup` pointer and then used a `SmallStack` that used this as the `Alloc` parameter?  Something like:

```c++
class BackupAlloc
{
  Block* backup = nullptr;
  Alloc *alloc;
  BackupAlloc(Alloc *a) : alloc(a) {}
  template<size_t Size>
  void * alloc()
  {
    static_assert(Size == sizeof(Block), "Allocating something not the size of a block");
    if (backup)
    {
      return std::exchange(backup, nullptr);
    }
    return alloc->template alloc<Size>();
  }
  template<size_t Size>
  void * dealloc(Block *b)
  {
    static_assert(Size == sizeof(Block), "Deallocating something not the size of a block");
    if (backup == nullptr)
    {
      backup = b;
    }
    else
    {
      alloc->template dealloc<Size>(b);
     }
  }
}
```

Reply:
This change would require the BackupAlloc to be passed each time.

I guess there are two aspects that differ between the two wrappers
* caching the alloc structure
* using a backup for a mini pool (this has a pretty big effect in some of the microbenchmarks for the runtime.

I'll see if I can pull the two aspects apart to make it clearer.

Reply:
With David's `BackupAlloc` you should be able to get rid of the distinction between StackSmall and StackBase.

`Stack<Alloc>` has a field `BackupAlloc<Alloc>` and `StackSmall<BackupAlloc<Alloc>>`, and it's implementation is limited to just augmenting the method calls with the extra BackupAlloc argument. I think that would also reduce the complexity of slow/fast path handling.

Reply:
You can't cache the Allocator if this is part of a region as it can be sent in a message to another thread, or an owning cown could move scheduler thread. So the reason for creating two types for the region tidy semantics we want, needs to be able to flow between scheduler threads.  The complexity was the only way I could see to share the code, and maintain the performance.

Reply:
This is what I had in mind: https://github.com/plietar/verona/commit/948431ee917da593582d1a0f43d3448d9f4dd7c3

As far as I can tell it behaves identically, and it keeps the control-flow more contralized.

Reply:
This is great, I tried something like this, but with the `snmalloc::function_ref` and it didn't work well.  Using the template works very nicely. Thanks

Reply:
Yup, @plietar's commit is exactly what I had in mind, sorry for not explaining it better.